File: automation-framework/utils/dataReader.ts
import fs from 'fs';
import path from 'path';

const TEST_DATA_DIR = path.join(__dirname, '..', 'test-data');

export function readJsonData<T>(fileName: string): T {
  return JSON.parse(fs.readFileSync(path.join(TEST_DATA_DIR, fileName), 'utf-8')) as T;
}

export function readCsvData(fileName: string): Record<string, string>[] {
  const lines = fs.readFileSync(path.join(TEST_DATA_DIR, fileName), 'utf-8').trim().split(/\r?\n/);
  const headers = lines[0].split(',').map((h) => h.trim());
  return lines.slice(1).map((line) => {
    const values = line.split(',').map((v) => v.trim());
    return headers.reduce<Record<string, string>>((row, header, i) => {
      row[header] = values[i] ?? '';
      return row;
    }, {});
  });
}

export interface TestUsersJson {
  validUser: { username: string; password: string };
  invalidPassword: { username: string; password: string };
  invalidUsername: { username: string; password: string };
  emptyCredentials: { username: string; password: string };
  products: { backpack: string; bikeLight: string };
}

export function getUsersFromJson(): TestUsersJson {
  return readJsonData<TestUsersJson>('users.json');
}

export type LoginScenarioKey = 'validUser' | 'invalidPassword' | 'invalidUsername' | 'emptyCredentials';

export function getLoginScenario(scenario: LoginScenarioKey) {
  return getUsersFromJson()[scenario];
}

export function getCsvScenario(scenario: string): Record<string, string> {
  const row = readCsvData('users.csv').find((item) => item.scenario === scenario);
  if (!row) throw new Error(`CSV scenario not found: ${scenario}`);
  return row;
}

export function getCheckoutCustomer() {
  return readJsonData<{ validCustomer: { firstName: string; lastName: string; postalCode: string } }>(
    'checkout_data.json',
  ).validCustomer;
}

File: ultis/data_reader.py
import csv
import json
from pathlib import Path

_PROJECT_ROOT = Path(__file__).resolve().parent.parent
_TEST_DATA_DIR = _PROJECT_ROOT / "test_data"


def load_json_data(file_name: str) -> dict:
    with open(_TEST_DATA_DIR / file_name, encoding="utf-8") as file:
        return json.load(file)


def load_csv_data(file_name: str):
    with open(_TEST_DATA_DIR / file_name, encoding="utf-8", newline="") as file:
        return list(csv.DictReader(file))


def get_login_scenario(scenario_name: str) -> dict:
    return load_json_data("login_data.json")[scenario_name]


def get_product_name(product_key: str) -> str:
    return load_json_data("product_data.json")[product_key]["name"]


def get_csv_scenario(scenario_name: str) -> dict:
    for row in load_csv_data("users.csv"):
        if row["scenario"] == scenario_name:
            return row
    raise KeyError(f"CSV scenario not found: {scenario_name}")


def get_checkout_customer() -> dict:
    return load_json_data("checkout_data.json")["valid_customer"]
